generator/dryrun.py

```diff
@@ -0,0 +1,273 @@
+                    job_config = bigquery.QueryJobConfig(
+                        dry_run=True,
+                        use_query_cache=False,
+                        query_parameters=[
+                            bigquery.ScalarQueryParameter(
+                                "submission_date", "DATE", "2019-01-01"
+                            )
+                        ],
+                    )
+                    job = self.client.query(self.sql, job_config=job_config)
+                    query_schema = (
+                        job._properties.get("statistics", {})
+                        .get("query", {})
+                        .get("schema", {})
+                    )
+                    referenced_tables = [
+                        ref.to_api_repr() for ref in job.referenced_tables
+                    ]
+
+                if (
+                    self.project is not None
+                    and self.table is not None
+                    and self.dataset is not None
+                ):
+                    table = self.client.get_table(
+                        f"{self.project}.{self.dataset}.{self.table}"
+                    )
+                    table_metadata = {
+                        "tableType": table.table_type,
+                        "friendlyName": table.friendly_name,
+                        "schema": {
+                            "fields": [field.to_api_repr() for field in table.schema]
+                        },
+                    }
+
+                return {
+                    "valid": True,
+                    "referencedTables": referenced_tables,
+                    "schema": query_schema,
+                    "tableMetadata": table_metadata,
+                }
+        except Exception as e:
+            print(f"ERROR {e}")
+            return None
+
+    def get_schema(self):
+        """Return the query schema by dry running the SQL file."""
+        if not self.is_valid():
+            raise DryRunError(
+                "Error when dry running SQL",
+                self.get_error(),
+                self.use_cloud_function,
+                self.table,
+            )
+
+        if (
+            self.dry_run_result
+            and self.dry_run_result["valid"]
+            and "schema" in self.dry_run_result
+        ):
+            return self.dry_run_result["schema"]["fields"]
+
+        return {}
+
+    def get_table_schema(self):
+        """Return the schema of the provided table."""
+        if not self.is_valid():
+            raise DryRunError(
+                "Error when dry running SQL",
+                self.get_error(),
+                self.use_cloud_function,
+                self.table,
+            )
+
+        if (
+            self.dry_run_result
+            and self.dry_run_result["valid"]
+            and "tableMetadata" in self.dry_run_result
+        ):
+            return self.dry_run_result["tableMetadata"]["schema"]["fields"]
+
+        return {}
```

Comment:
Shouldn't this be an empty list?
```suggestion
        return []
```